halo2_proofs/src/arithmetic.rs

```diff
@@ -189,79 +178,107 @@ fn serial_fft<G: Group>(a: &mut [G], omega: G::Scalar, log_n: u32) {
         r
     }
 
-    let n = a.len() as u32;
+    fn log2_floor() -> u32 {
+        let num = multicore::current_num_threads();
+        assert!(num > 0);
+        let mut pow = 0;
+        while (1 << (pow + 1)) <= num {
+            pow += 1;
+        }
+        pow
+    }
+
+    let log_threads = log2_floor();
+    let n = a.len() as usize;
     assert_eq!(n, 1 << log_n);
 
     for k in 0..n {
-        let rk = bitreverse(k, log_n);
+        let rk = bitreverse(k, log_n as usize);
         if k < rk {
-            a.swap(rk as usize, k as usize);
+            a.swap(rk, k);
         }
     }
 
-    let mut m = 1;
-    for _ in 0..log_n {
-        let w_m = omega.pow_vartime(&[u64::from(n / (2 * m)), 0, 0, 0]);
-
-        let mut k = 0;
-        while k < n {
-            let mut w = G::Scalar::one();
-            for j in 0..m {
-                let mut t = a[(k + j + m) as usize];
-                t.group_scale(&w);
-                a[(k + j + m) as usize] = a[(k + j) as usize];
-                a[(k + j + m) as usize].group_sub(&t);
-                a[(k + j) as usize].group_add(&t);
-                w *= &w_m;
-            }
-
-            k += 2 * m;
-        }
+    // precompute twiddle factors
+    let mut w = G::Scalar::one();
+    let mut twiddles = vec![G::Scalar::one(); (n / 2) as usize];
 
-        m *= 2;
+    for tw in twiddles.iter_mut() {
+        *tw = w;
+        w.group_scale(&omega);
     }
-}
-
-fn parallel_fft<G: Group>(a: &mut [G], omega: G::Scalar, log_n: u32, log_threads: u32) {
-    assert!(log_n >= log_threads);
-
-    let num_threads = 1 << log_threads;
-    let log_new_n = log_n - log_threads;
-    let mut tmp = vec![vec![G::group_zero(); 1 << log_new_n]; num_threads];
-    let new_omega = omega.pow_vartime(&[num_threads as u64, 0, 0, 0]);
-
-    multicore::scope(|scope| {
-        let a = &*a;
 
-        for (j, tmp) in tmp.iter_mut().enumerate() {
-            scope.spawn(move |_| {
-                // Shuffle into a sub-FFT
-                let omega_j = omega.pow_vartime(&[j as u64, 0, 0, 0]);
-                let omega_step = omega.pow_vartime(&[(j as u64) << log_new_n, 0, 0, 0]);
-
-                let mut elt = G::Scalar::one();
-
-                for (i, tmp) in tmp.iter_mut().enumerate() {
-                    for s in 0..num_threads {
-                        let idx = (i + (s << log_new_n)) % (1 << log_n);
-                        let mut t = a[idx];
-                        t.group_scale(&elt);
-                        tmp.group_add(&t);
-                        elt *= &omega_step;
-                    }
-                    elt *= &omega_j;
-                }
-
-                // Perform sub-FFT
-                serial_fft(tmp, new_omega, log_new_n);
+    if log_n <= log_threads {
```

Comment:
This uses a non-recursive implementation for the whole FFT when log_n ≤ log_threads. If it is not, then the recursive butterfly recurses all the way down to n = 2. My intuition would be that it's best to bottom out with the non-recursive implementation at a higher n, i.e. to switch from recursive to non-recursive when n becomes small enough within a large FFT.

Reply:
> My intuition would be that it's best to bottom out at a higher n

Yes.
The computation doesn't cost much so the less multiplicity, the faster butterfly performs.
> to switch from recursive to non-recursive when n becomes small enough within a large FFT

I think all recursive is the best option because of rayon specification.
We use `join` for recursive and `scope` for for loop.
The `join` allocates stack and `scope` allocates heap.
We should use `join` as possible as we can as the [experiment](https://github.com/appliedzkp/halo2/pull/55#issue-1195660345) suggests.

I think we can improve more to implement following.

- implement 2 layer butterfly arithmetic

As you suggest, the less multiplicity, the faster butterfly performs.
If we perform 2 layer butterfly arithmetic, it would improve performance and be able to user intermediate value.
But code will be mess as following.
https://github.com/appliedzkp/halo2/blob/fft-optimization/halo2_proofs/src/arithmetic.rs#L216
It's not so hard just adding `if n == 4` condition and divide by 4.
Should I implement it?

- change `par_iter` when the n is large enough

I use `left.iter_mut` when we perform butterfly arithmetic in `recursive_butterfly`.
https://github.com/zcash/halo2/pull/543/files#diff-e03420d624b59ef5f1ea16419344db88889e1864d7b089db7653ae943b4c7809R272
We can replace with `par_iter_mut` but it was bottleneck because the overhead of multithread costs more than speed up.
If we replace `iter_mut` with `par_iter_mut` at proper n, it would improve performance but we shouldn't use `if` condition because it slows down.
But we need some experiments.

Reply:
I don't find this code too bad: https://github.com/appliedzkp/halo2/blob/fft-optimization/halo2_proofs/src/arithmetic.rs#L216
It seems feasibly reviewable and would be worth the complexity if it gave, say, another 5% improvement or more.

Reply:
Copy that.